lib/xdrgen/generators/go.rb

```diff
@@ -336,6 +419,126 @@ def render_binary_interface(out, name)
         out.break
       end
 
+      # render_encode_to_body assumes there is an `e` variable containing an
+      # xdr.Encoder, and a variable defined by `name` that is the value to
+      # encode.
+      def render_encode_to_body(out, var, type, self_encode:)
+        optional = type.sub_type == :optional
+        if optional
+          out.puts "  _, err = e.EncodeBool(#{var} != nil)"
+          out.puts "  if err != nil {"
+          out.puts "    return err"
+          out.puts "  }"
+          out.puts "  if #{var} != nil {"
+          var = "(*#{var})"
+        end
+        case type
+        when AST::Typespecs::UnsignedHyper
+          out.puts "  _, err = e.EncodeUhyper(uint64(#{var}))"
+        when AST::Typespecs::Hyper
+          out.puts "  _, err = e.EncodeHyper(int64(#{var}))"
+        when AST::Typespecs::UnsignedInt
+          out.puts "  _, err = e.EncodeUint(uint32(#{var}))"
+        when AST::Typespecs::Int, AST::Definitions::Enum
+          out.puts "  _, err = e.EncodeInt(int32(#{var}))"
+        when AST::Typespecs::String
+          out.puts "  _, err = e.EncodeString(string(#{var}))"
+        when AST::Typespecs::Opaque
+          if type.fixed?
+            out.puts "  _, err = e.EncodeFixedOpaque(#{var}[:])"
+          else
+            out.puts "  _, err = e.EncodeOpaque(#{var}[:])"
+          end
+        when AST::Typespecs::Simple
+          case type.sub_type
+          when :simple, :optional
+            optional_within = type.is_a?(AST::Identifier) && type.resolved_type.sub_type == :optional
+            if optional_within
+              out.puts "  _, err = e.EncodeBool(#{var} != nil)"
+              out.puts "  if err != nil {"
+              out.puts "    return err"
+              out.puts "  }"
+              out.puts "  if #{var} != nil {"
+              var = "(*#{var})"
+            end
+            var = "#{name type}(#{var})" if self_encode
+            out.puts "  err = #{var}.EncodeTo(e)"
+            if optional_within
+              out.puts "  }"
+            end
+          when :array
+            out.puts "  for i := 0; i < len(#{var}); i++ {"
```

Comment:
I don't think it is necessary because the type of `var` in the `:array` case is a fixed size array, and in Go we will represent it as an array, which has a fixed size at compile time. For example:

https://github.com/stellar/go/blob/a1db2a6b1fd3a444cdc9ac7250144e28d5de5354/xdr/xdr_generated.go#L18912-L18916

However, we could check max_length for the `:var_array` case which we represent as a Go slice and therefore has an unbounded size. The code that does encoding of variable length arrays today (src: https://github.com/stellar/go-xdr/blob/b95df30963cd0a4fcd12f2d65db0f9aeba987f73/xdr3/encode.go#L404-L427) doesn't check the max length, so that would be a functional change. I think if we are to pursue that we should pursue that separately to this PR.